### PR TITLE
[aztables] adding CosmosDB support for using TokenCredentials

### PR DESCRIPTION
**Pull request by richardpark-msft**
CosmosDB support for using TokenCredentials, including a lot of test rework to bring it up to date with our standards.

### CODE REVIEW

File: eng/config.json [hunk 1]
```diff
@@ -72,6 +72,15 @@
             "Name": "keyvault/azsecrets",
             "CoverageGoal": 0.86
         },
+        {
+            "Name": "data/aztables",
+            "CoverageGoal": 0.77,
+            "EnvironmentVariables": {
+                "TABLES_STORAGE_ACCOUNT_NAME": "fakeaccount",
+                "TABLES_PRIMARY_STORAGE_ACCOUNT_KEY": "Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/KBHBeksoGMGw==",
+                "TABLES_SHARED_ACCESS_SIGNATURE": "?sig=Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/KBHBeksoGMGw=="
```
Comment (RickWinter) [inline eng/config.json:81]: Are these values going to get flagged by cred scanners?

File: sdk/data/aztables/README.md [hunk 2]
```diff
@@ -59,9 +59,8 @@ func main() {
 For more information about table service URL's and how to configure custom domain names for Azure Storage check out the [official documentation][azure_portal_account_url]
 
 #### Types of credentials
-The clients support different forms of authentication. Cosmos accounts can use a Shared Key Credential, Connection String, or an Shared Access Signature Token for authentication. Storage account can use the same credentials as a Cosmos account and can use the credentials in [`azidentity`](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/azidentity) like `azidentity.NewDefaultAzureCredential()`.
 
-The aztables package supports any of the types that implement the `azcore.TokenCredential` interface, authorization via a Connection String, or authorization with a Shared Access Signature Token.
+Both services (Cosmos and Storage) support the the following forms of authentication: a Shared Key Credential, Connection String, Shared Access Signature Token or a credential from the [`azidentity`](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/azidentity) module, like `azidentity.DefaultAzureCredential`.
```
Comment (RickWinter) [inline sdk/data/aztables/README.md:63]: Reword this to have the identity credentials first. Customers should use those. We should downplay the other options.

File: eng/config.json (hunk 1 above)
Comment (richardpark-msft) [reply eng/config.json:81]: It's not new, they were part of the previous 'data' block.

File: sdk/data/aztables/service_client.go [hunk 3]
```diff
@@ -16,6 +16,9 @@ import (
 	generated "github.com/Azure/azure-sdk-for-go/sdk/data/aztables/internal"
 )
 
+var storageScope = []string{"https://storage.azure.com/.default"}
```
[the remark this answers is not in the document]
Comment (richardpark-msft) [reply sdk/data/aztables/service_client.go]: They're only used in this file, AFAICT but a 'var' block is nicer looking anyways.

File: sdk/data/aztables/README.md (hunk 2 above)
Comment (richardpark-msft) [reply sdk/data/aztables/README.md]: Updated.

File: eng/config.json (hunk 1 above)
Comment (chlowell) [reply eng/config.json:81]: Can we use a more obviously fake value everywhere?

**Comment by richardpark-msft** [comment]
All done with feedback if someone wants to approve.

File: sdk/data/aztables/proxy_test.go [hunk 4]
```diff
@@ -16,10 +16,12 @@ import (
 	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
 	"github.com/Azure/azure-sdk-for-go/sdk/azcore/tracing"
 	"github.com/Azure/azure-sdk-for-go/sdk/internal/recording"
+	"github.com/Azure/azure-sdk-for-go/sdk/internal/test/credential"
 	"github.com/stretchr/testify/require"
 )
 
 const recordingDirectory = "sdk/data/aztables/testdata"
+const fakeAccount = "Sanitized"
```
Comment (chlowell) [inline sdk/data/aztables/proxy_test.go:24]:
```suggestion
const fakeAccount = recording.SanitizedValue
```

File: sdk/data/aztables/proxy_test.go [hunk 5]
```diff
@@ -149,121 +192,169 @@ func initClientTest(t *testing.T, service string, createTable bool, tp tracing.P
...
 	serviceURL := storageURI(accountName)
 
 	tableName, err := createRandomName(t, tableNamePrefix)
 	require.NoError(t, err)
 
-	return createClientForRecording(t, tableName, serviceURL, *cred, tp)
+	if options.UseSharedKey {
+		var cred *SharedKeyCredential
+
+		if recording.GetRecordMode() == "playback" {
```
Comment (chlowell) [inline sdk/data/aztables/proxy_test.go:261]:
```suggestion
		if recording.GetRecordMode() == recording.PlaybackMode {
```

File: sdk/data/aztables/proxy_test.go [hunk 6]
```diff
@@ -149,121 +192,169 @@ func initClientTest(t *testing.T, service string, createTable bool, tp tracing.P
...
+
+func createCosmosClient(t *testing.T, tp tracing.Provider, options *testClientOptions) (*Client, error) {
+	if options == nil {
+		options = &testClientOptions{}
 	}
 
-	cred, err := getSharedKeyCredential()
-	require.NoError(t, err)
+	accountName := recording.GetEnvVariable("TABLES_COSMOS_ACCOUNT_NAME", fakeAccount)
+	if recording.GetRecordMode() == "playback" {
```
Comment (chlowell) [inline sdk/data/aztables/proxy_test.go:285]:
```suggestion
	if recording.GetRecordMode() == recording.PlaybackMode {
```

File: sdk/data/aztables/proxy_test.go [hunk 7]
```diff
@@ -149,121 +192,169 @@ func initClientTest(t *testing.T, service string, createTable bool, tp tracing.P
...
+	accountName := recording.GetEnvVariable("TABLES_STORAGE_ACCOUNT_NAME", fakeAccount)
+	accountKey := recording.GetEnvVariable("TABLES_PRIMARY_STORAGE_ACCOUNT_KEY", "fakeaccountkey")
 	serviceURL := storageURI(accountName)
 
-	return createServiceClientForRecording(t, serviceURL, *cred, tp)
+	if options.UseSharedKey {
+		var cred *SharedKeyCredential
+		var err error
+
+		if recording.GetRecordMode() == "playback" {
```
Comment (chlowell) [inline sdk/data/aztables/proxy_test.go:316]:
```suggestion
		if recording.GetRecordMode() == recording.PlaybackMode {
```

File: sdk/data/aztables/proxy_test.go [hunk 8]
```diff
@@ -149,121 +192,169 @@ func initClientTest(t *testing.T, service string, createTable bool, tp tracing.P
...
+func createCosmosServiceClient(t *testing.T, tp tracing.Provider, options *testClientOptions) (*ServiceClient, error) {
+	if options == nil {
+		options = &testClientOptions{}
 	}
 
-	cred, err := getSharedKeyCredential()
-	require.NoError(t, err)
+	accountName := recording.GetEnvVariable("TABLES_COSMOS_ACCOUNT_NAME", fakeAccount)
+
+	if recording.GetRecordMode() == "playback" {
```
Comment (chlowell) [inline sdk/data/aztables/proxy_test.go:337]:
```suggestion
	if recording.GetRecordMode() == recording.PlaybackMode {
```

File: sdk/data/aztables/test-resources.bicep [hunk 9]
```diff
@@ -0,0 +1,158 @@
+param baseName string
```
Comment (chlowell) [inline sdk/data/aztables/test-resources.bicep:1]:
```suggestion
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

param baseName string
```

File: sdk/data/aztables/testdata/perf/go.mod [hunk 10]
```diff
@@ -3,14 +3,14 @@ module github.com/Azure/azure-sdk-for-go/sdk/data/aztables/testdata/perf
 go 1.18
 
 require (
-	github.com/Azure/azure-sdk-for-go/sdk/azcore v1.13.0
+	github.com/Azure/azure-sdk-for-go/sdk/azcore v1.15.0
```
Comment (chlowell) [inline sdk/data/aztables/testdata/perf/go.mod:6]:
```suggestion
	github.com/Azure/azure-sdk-for-go/sdk/azcore v1.16.0
```
😁

**Review by chlowell — APPROVED** [review]

File: eng/config.json (hunk 1 above)
Comment (richardpark-msft) [reply eng/config.json:81]: Ah, these are Azurite (ie, the emulator) connection values. So they're not secret at all.

**Comment by richardpark-msft** [comment]
@jhendrixMSFT, you (and now I) are the only CODEOWNERs for this library so I need you to approve.

**Comment by richardpark-msft** [comment]
@benbp or @weshaggard, I believe you have to approve this as well since I altered CODEOWNERS.

**Review by jhendrixMSFT — APPROVED** [review]

File: sdk/data/aztables/test-resources.bicep [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+    file: {
+      enabled: true
+    }
+    blob: {
+      enabled: true
+    }
+  }
+  keySource: 'Microsoft.Storage'
+}
+var networkAcls = {
```
Comment (benbp) [inline sdk/data/aztables/test-resources.bicep:26]: Are these necessary or copied? Ideally we don't have allow set as default. If possible I would remove the whole net acl block.

Comment (benbp) [reply sdk/data/aztables/test-resources.bicep:26]: Oh this is a straight conversion of existing, nvm

**Review by benbp — APPROVED** [review]

File: sdk/data/aztables/test-resources.bicep (hunk 11 above)
Comment (richardpark-msft) [reply sdk/data/aztables/test-resources.bicep:26]: Yeah, should mention, after talking with Ben, that these apparently get injected into the CI pipeline. There are other ways to do this, and they're probably more self-explanatory. Will do this in a separate PR.

Comment (richardpark-msft) [reply sdk/data/aztables/test-resources.bicep:26]: Issue filed: https://github.com/Azure/azure-sdk-for-go/issues/23603
